File: fe/fe-core/src/main/java/org/apache/doris/common/plugin/PluginFileCache.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.doris.common.plugin;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import com.google.common.base.Strings;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.File;

/**
 * PluginFileCache - Simple static memory cache for plugin file validation.
 * <p>
 * Caching logic:
 * 1. Users do not provide MD5 -> Use it directly if there are files locally (most efficient)
 * 2. User provides MD5 -> Quick check using cached MD5 (avoid recalculation)
 */
public class PluginFileCache {
    private static final Logger LOG = LogManager.getLogger(PluginFileCache.class);

    // Caffeine LRU cache with max 100 entries
    private static final Cache<String, FileInfo> CACHE = Caffeine.newBuilder()
            .maximumSize(100)
            .build();

    /**
     * Simple file info for caching
     */
    static class FileInfo {
        String localMd5;        // The MD5 value of the file for quick verification
        long fileSize;          // File size

        FileInfo() {
        }
    }

    /**
     * Check if local file is valid based on user requirements
     *
     * @param localPath local file path
     * @param userMd5 user provided MD5, can be null
     * @return true if file is valid and no download needed
     */
    public static boolean isFileValid(String localPath, String userMd5) {
        try {
            File file = new File(localPath);
            if (!file.exists() || file.length() == 0) {
                return false; // Case: No files -> download required
            }
            // Case: The user provides MD5 -> must validate the MD5 of the local cache
            if (!Strings.isNullOrEmpty(userMd5)) {
                FileInfo info = CACHE.getIfPresent(localPath);
                if (info == null || Strings.isNullOrEmpty(info.localMd5)) {
                    // There is no MD5 information in the cache, so you need to download and verify it again
                    return false;
                }
                // Use cached MD5 for quick verification
                return userMd5.equalsIgnoreCase(info.localMd5);
            }
            // Case: Users do not provide MD5 -> Use it directly if you have a file
            return true;
        } catch (Exception e) {
            LOG.warn("Failed to validate file {}: {}", localPath, e.getMessage());
            return false;
        }
    }

    /**
     * Update cache after successful download
     *
     * @param localPath local file path
     * @param localMd5 MD5 hash of the local file
     * @param fileSize size of the file in bytes
     */
    public static void updateCache(String localPath, String localMd5, long fileSize) {
        try {
            File file = new File(localPath);
            if (!file.exists()) {
                return;
            }
            FileInfo info = new FileInfo();
            info.localMd5 = localMd5;
            info.fileSize = fileSize;
            CACHE.put(localPath, info);
        } catch (Exception e) {
            LOG.warn("Failed to update cache for {}: {}", localPath, e.getMessage());
        }
    }
}

File: be/src/runtime/plugin/plugin_file_cache.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "runtime/plugin/plugin_file_cache.h"

#include <filesystem>
#include <string>

#include "common/logging.h"

namespace doris {

// Static member definitions
LruCache<std::string, PluginFileCache::FileInfo> PluginFileCache::_cache {MAX_CACHE_SIZE};
std::mutex PluginFileCache::_cache_mutex;

bool PluginFileCache::is_file_valid(const std::string& local_path, const std::string& user_md5) {
    try {
        // Check if file exists and is not empty
        if (!std::filesystem::exists(local_path) || std::filesystem::file_size(local_path) == 0) {
            return false; // Case: No files -> download required
        }

        // Case: User provides MD5 -> must validate the MD5 of the local cache
        if (!user_md5.empty()) {
            std::lock_guard lock(_cache_mutex);
            FileInfo cached_info;
            if (_cache.get(local_path, &cached_info)) {
                // Found in cache, use cached MD5 for quick verification
                return user_md5 == cached_info.local_md5;
            }
            // Not in cache, need to download and verify
            return false;
        }

        // Case: User does not provide MD5 -> Use it directly if you have a file (most efficient)
        return true;

    } catch (const std::exception& e) {
        LOG(WARNING) << "Failed to validate file " << local_path << ": " << e.what();
        return false;
    }
}

void PluginFileCache::update_cache(const std::string& local_path, const std::string& local_md5,
                                   long file_size) {
    try {
        if (!std::filesystem::exists(local_path)) {
            return;
        }
        std::lock_guard lock(_cache_mutex);
        _cache.put(local_path, FileInfo(local_md5, file_size));
    } catch (const std::exception& e) {
        LOG(WARNING) << "Failed to update cache for " << local_path << ": " << e.what();
    }
}

} // namespace doris